File: misc/200.number_of_islands.py
from typing import List
    
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        islands = 0
        for i, row in enumerate(grid):
            for j, cell in enumerate(row):
                if cell == '1':
                    islands += 1
                    self.remove_island(grid, i, j)
        return islands

    def remove_island(self, grid: List[List[str]], i: int, j: int) -> None:
        # Check out of bounds and not water
        if (i < 0) or (j < 0) or (i >= len(grid)) or (j >= len(grid[0])) or grid[i][j] == '0':
            return;
        grid[i][j] = '0' # Invalidate
        self.remove_island(grid, i, j - 1)  # left
        self.remove_island(grid, i, j + 1)  # right
        self.remove_island(grid, i - 1, j)  # up
        self.remove_island(grid, i + 1, j)  # down
        return 

File: pres/200.number_of_islands.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int islands = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    islands += 1;
                    remove_islands(grid, i, j, m, n);
                }
            }
        }
        return islands;
    }
    void remove_islands(vector<vector<char>>& grid, int i, int j, int m, int n) {
        // Check bounds and not water
        if (i < 0 || j < 0 || i >= m || j >= n || grid[i][j] == '0')
            return;

        grid[i][j] = '0';
        remove_islands(grid, i, j - 1, m, n);   // left
        remove_islands(grid, i, j + 1, m, n);   // right
        remove_islands(grid, i - 1, j, m, n);   // up
        remove_islands(grid, i + 1, j, m, n);   // down
    }
};
